File: examples/ledger/elixir/test/ledger_test.exs
defmodule LedgerTest do
  use ExUnit.Case

  alias BowlineClient.Error
  alias Ledger.Invoices
  alias Ledger.Types

  setup_all do
    port = free_port()
    ledger_dir = Path.expand("../..", __DIR__)

    server =
      Port.open({:spawn_executable, System.find_executable("go")}, [
        :binary,
        :exit_status,
        args: ["run", "./cmd/server"],
        cd: ledger_dir,
        env: [
          {~c"ADDR", ~c"127.0.0.1:#{port}"},
          {~c"LEDGER_FIXED_TIME", ~c"2026-09-15T12:00:00Z"}
        ]
      ])

    {:os_pid, os_pid} = Port.info(server, :os_pid)
    base_url = "http://127.0.0.1:#{port}/api"
    wait_for_health(base_url, 300)

    on_exit(fn ->
      System.cmd("pkill", ["-P", Integer.to_string(os_pid)], stderr_to_stdout: true)
      System.cmd("kill", [Integer.to_string(os_pid)], stderr_to_stdout: true)
    end)

    {:ok, transport: Ledger.Client.new(base_url)}
  end

  test "lists the two seeded invoices", %{transport: transport} do
    assert {:ok, %Types.Page{items: items}} =
             Invoices.list(transport, %Types.ListInvoicesInput{limit: 20})

    seeded = Enum.filter(items, &(&1.id in [3, 4]))
    assert Enum.map(seeded, & &1.id) == [3, 4]
    assert [%Types.Invoice{status: :sent, total: "USD 1500.00"} | _] = seeded
    assert %DateTime{} = hd(seeded).created_at
  end

  test "a bad create surfaces the server's validation issues", %{transport: transport} do
    input = %Types.CreateInvoiceInput{
      customer_id: 1,
      lines: [%Types.Line{description: "", quantity: 0, unit_price: "USD 10.00"}]
    }

    assert {:error, %Error{code: :invalid_argument, status: 400, issues: issues}} =
             Invoices.create(transport, input)

    assert Enum.map(issues, & &1.path) == [
             ["lines", "0", "description"],
             ["lines", "0", "quantity"]
           ]

    assert Types.CreateInvoiceInput.validate(input) |> Enum.map(& &1.path) ==
             Enum.map(issues, & &1.path)
  end

  test "creates, watches, and voids an invoice", %{transport: transport} do
    task =
      Task.async(fn ->
        {:ok, stream} = Invoices.watch(transport, %Types.WatchInput{})
        stream |> Enum.take(1) |> hd()
      end)

    Process.sleep(200)

    input = %Types.CreateInvoiceInput{
      customer_id: 1,
      lines: [%Types.Line{description: "Widgets", quantity: 3, unit_price: "USD 10.00"}],
      note: "net 30"
    }

    assert {:ok, %Types.Invoice{} = created} = Invoices.create(transport, input)
    assert created.total == "USD 30.00"
    assert created.status == :draft
    assert created.note == "net 30"

    assert %Types.Invoice{id: watched_id} = Task.await(task, 5_000)
    assert watched_id == created.id

    assert {:ok, %Types.Invoice{status: :void}} =
             Invoices.void(transport, %Types.VoidInvoiceInput{id: created.id})

    assert {:error, %Error{code: :failed_precondition, type: "InvoiceLocked"} = error} =
             Invoices.void(transport, %Types.VoidInvoiceInput{id: 4})

    assert %Types.InvoiceLocked{id: 4, status: :paid} = Ledger.Client.error_details(error)
  end

  defp free_port do
    {:ok, socket} = :gen_tcp.listen(0, [])
    {:ok, port} = :inet.port(socket)
    :gen_tcp.close(socket)
    port
  end

  defp wait_for_health(_base_url, 0), do: raise("the ledger server did not start")

  defp wait_for_health(base_url, attempts) do
    case Req.get(base_url <> "/health", retry: false) do
      {:ok, %Req.Response{status: 200}} ->
        :ok

      _ ->
        Process.sleep(200)
        wait_for_health(base_url, attempts - 1)
    end
  end
end

File: examples/ledger/dart/test/ledger_test.dart
import 'dart:async';
import 'dart:io';

import 'package:bowline/bowline.dart';
import 'package:http/http.dart' as http;
import 'package:ledger_dart/bowline.dart';
import 'package:test/test.dart';

Future<int> freePort() async {
  final socket = await ServerSocket.bind(InternetAddress.loopbackIPv4, 0);
  final port = socket.port;
  await socket.close();
  return port;
}

Future<void> waitForHealth(Uri url) async {
  final deadline = DateTime.now().add(const Duration(seconds: 90));
  while (DateTime.now().isBefore(deadline)) {
    try {
      final response = await http.get(url);
      if (response.statusCode == 200) {
        return;
      }
    } on http.ClientException {
      await Future<void>.delayed(const Duration(milliseconds: 200));
    } on SocketException {
      await Future<void>.delayed(const Duration(milliseconds: 200));
    }
  }
  throw StateError('ledger server did not become healthy');
}

void main() {
  late Process server;
  late Client client;
  late Uri base;

  setUpAll(() async {
    final port = await freePort();
    server = await Process.start(
      'go',
      ['run', './cmd/server'],
      workingDirectory: '..',
      environment: {'ADDR': '127.0.0.1:$port'},
    );
    server.stdout.drain<void>();
    server.stderr.drain<void>();
    base = Uri.parse('http://127.0.0.1:$port/api');
    await waitForHealth(base.replace(path: '${base.path}/health'));
    client = Client(Transport(baseUrl: base));
  });

  tearDownAll(() async {
    server.kill();
    await server.exitCode.timeout(const Duration(seconds: 10), onTimeout: () {
      server.kill(ProcessSignal.sigkill);
      return -1;
    });
  });

  test('lists the seeded invoices', () async {
    final page = await client.invoices.list(ListInvoicesInput(limit: 20));
    final ids = page.items.map((i) => i.id).toList();
    expect(ids, containsAll([3, 4]));
    final paid = page.items.firstWhere((i) => i.id == 4);
    expect(paid.status, Status.paid);
    expect(paid.total, 'USD 9999.00');
    expect(paid.createdAt.isUtc, isTrue);
  });

  test('validation issues surface locally with server paths', () async {
    try {
      await client.invoices.create(CreateInvoiceInput(
        customerId: 1,
        lines: [Line(description: '', quantity: 0, unitPrice: 'USD 1.00')],
      ));
      fail('expected a validation failure');
    } on BowlineException catch (e) {
      expect(e.code, Code.invalidArgument);
      expect(e.issues.map((i) => i.path.join('.')),
          containsAll(['lines.0.description', 'lines.0.quantity']));
    }
  });

  test('the server rejects an input the client did not check', () async {
    final raw = Transport(baseUrl: base);
    try {
      await raw.call(
          'invoices.create',
          Method.post,
          {
            'customerId': 1,
            'lines': [
              {'description': '', 'quantity': 0, 'unitPrice': 'USD 1.00'}
            ]
          },
          asRaw);
      fail('expected a validation failure');
    } on BowlineException catch (e) {
      expect(e.status, 400);
      expect(e.issues.map((i) => i.path.join('.')),
          containsAll(['lines.0.description', 'lines.0.quantity']));
    }
  });

  test('creates, watches, and voids', () async {
    final events = <Invoice>[];
    final subscription = client.invoices
        .watch(WatchInput())
        .listen(events.add, onError: (Object e) {});
    await Future<void>.delayed(const Duration(milliseconds: 300));
    final created = await client.invoices.create(CreateInvoiceInput(
      customerId: 1,
      lines: [
        Line(description: 'Widgets', quantity: 3, unitPrice: 'USD 10.00')
      ],
      note: 'from dart',
    ));
    expect(created.total, 'USD 30.00');
    expect(created.note, 'from dart');
    final voided =
        await client.invoices.void_(VoidInvoiceInput(id: created.id));
    expect(voided.status, Status.void_);
    await Future<void>.delayed(const Duration(milliseconds: 300));
    await subscription.cancel();
    expect(events.map((e) => e.id), contains(created.id));
  });

  test('a locked invoice surfaces FAILED_PRECONDITION with typed details',
      () async {
    try {
      await client.invoices.void_(VoidInvoiceInput(id: 4));
      fail('expected InvoiceLocked');
    } on BowlineException catch (e) {
      expect(e.code, Code.failedPrecondition);
      expect(e.status, 412);
      expect(e.type, 'InvoiceLocked');
      final details = InvoiceLocked.fromJson(asObject(e.details));
      expect(details.id, 4);
      expect(details.status, Status.paid);
    }
  });
}
